overlord/hookstate/hookmgr.go

```diff
@@ -0,0 +1,159 @@
+}
+
+// HandlerGenerator is the function signature required to register for hooks.
+type HandlerGenerator func(*Context) Handler
+
+// hookSetup is a reference to a hook within a specific snap.
+type hookSetup struct {
+	Snap     string        `json:"snap"`
+	Revision snap.Revision `json:"revision"`
+	Hook     string        `json:"hook"`
+}
+
+// Manager returns a new HookManager.
+func Manager(s *state.State) (*HookManager, error) {
+	runner := state.NewTaskRunner(s)
+	manager := &HookManager{
+		state:      s,
+		runner:     runner,
+		repository: newRepository(),
+	}
+
+	runner.AddHandler("run-hook", manager.doRunHook, nil)
+
+	return manager, nil
+}
+
+// HookTask returns a task that will run the specified hook.
+func HookTask(s *state.State, taskSummary, snapName string, revision snap.Revision, hookName string) *state.Task {
+	task := s.NewTask("run-hook", taskSummary)
+	task.Set(hookSetupKey, hookSetup{Snap: snapName, Revision: revision, Hook: hookName})
+	return task
+}
+
+// Register requests that a given handler generator be called when a matching
+// hook is run, and the handler be used for the hook.
+//
+// Specifically, if a matching hook is about to be run, the handler's Before()
+// method will be called. After the hook has completed running, either the
+// handler's Done() method or its Error() method will be called, depending on
+// the outcome.
+func (m *HookManager) Register(pattern *regexp.Regexp, generator HandlerGenerator) {
+	m.repository.addHandlerGenerator(pattern, generator)
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *HookManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *HookManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *HookManager) Stop() {
+	m.runner.Stop()
+}
+
+// doRunHook actually runs the hook that was requested.
+//
+// Note that this method is synchronous, as the task is already running in a
+// goroutine.
+func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
+	task.State().Lock()
+	var setup hookSetup
+	err := task.Get(hookSetupKey, &setup)
+	task.State().Unlock()
+
+	if err != nil {
+		return fmt.Errorf("cannot extract hook setup from task: %s", err)
+	}
+
+	// Obtain a handler for this hook. The repository returns a list since it's
+	// possible for regular expressions to overlap, but multiple handlers is an
+	// error (as is no handler).
+	handlers := m.repository.generateHandlers(newContext(task, setup))
+	handlersCount := len(handlers)
+	if handlersCount == 0 {
+		return fmt.Errorf("no registered handlers for hook %q", setup.Hook)
+	}
+	if handlersCount > 1 {
+		return fmt.Errorf("%d handlers registered for hook %q, expected 1", handlersCount, setup.Hook)
+	}
+
+	handler := handlers[0]
+
+	// About to run the hook-- notify the handler
+	if err := handler.Before(); err != nil {
+		return fmt.Errorf("cannot handle hook %q: %s", setup.Hook, err)
```

Comment:
I think we should just pass the error up here. The failure happened while handling the hook, so we could handle it, and we don't quite know why the handler decided to error out. But it's our code, and this means it should have a good error message which we can just pass on. Also, it's worth noting that this error message will be in the context of the task summary, which will have the hook details.